File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        //windowEnd - windowStart + 1 -> answer - > to be maximized
        int windowStart = 0, max = Integer.MIN_VALUE, maxRepeatingChar = 0;
        Map<Character, Integer> map = new HashMap<>();
        for (int windowEnd = 0; windowEnd < s.length(); windowEnd++){
            char curr = s.charAt(windowEnd);
            map.put(curr, map.getOrDefault(curr, 0) +1);
            maxRepeatingChar = Math.max(maxRepeatingChar, map.get(curr));
            int windowSize = windowEnd - windowStart + 1;
            if (windowSize - maxRepeatingChar > k){
                char leftChar = s.charAt(windowStart);
                map.put(leftChar, map.get(leftChar) - 1);
                windowStart++;
            }
            max = Math.max(max, windowEnd - windowStart + 1);
        }
        
        return max;
    }
}

File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        window_start = max_len = max_repeat_char = 0
        char_freq_map = {}
        for window_end in range(len(s)):
            right_char = s[window_end]
            
            if right_char not in char_freq_map:
                char_freq_map[right_char] = 0
            char_freq_map[right_char] += 1
            
            max_repeat_char = max(max_repeat_char, char_freq_map[right_char])
            
            window_size = window_end - window_start + 1
            if (window_size - max_repeat_char) > k:
                left_char = s[window_start]
                char_freq_map[left_char] -= 1
                window_start += 1
                
            max_len = max(max_len, window_end - window_start + 1)
        
        return max_len